README.md
# elenpi-html-parser

[![Travis branch](https://img.shields.io/travis/nomocas/elenpi-html-parser/master.svg)](https://travis-ci.org/nomocas/elenpi-html-parser)
[![bitHound Overall Score](https://www.bithound.io/github/nomocas/elenpi-html-parser/badges/score.svg)](https://www.bithound.io/github/nomocas/elenpi-html-parser)
[![bitHound Dependencies](https://www.bithound.io/github/nomocas/elenpi-html-parser/badges/dependencies.svg)](https://www.bithound.io/github/nomocas/elenpi-html-parser/master/dependencies/npm)
[![bitHound Dev Dependencies](https://www.bithound.io/github/nomocas/elenpi-html-parser/badges/devDependencies.svg)](https://www.bithound.io/github/nomocas/elenpi-html-parser/master/dependencies/npm)
[![licence](https://img.shields.io/npm/l/elenpi-html-parser.svg)](https://spdx.org/licenses/MIT)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)

The simplest, lightest, most understandable and tweakable HTML5 parser.

Not the fastest but quite fast as well.

Work browser side and server side.

## Usage


```shell
npm i elenpi-html-parser --save
```

```javascript
const parser = require('elenpi-html-parser');

const fragmentAst = parser.parseFragment('<div class="foo">hello world</div>');

const docAST = parser.parseFragment(`
<!doctype html>
<html>
<head>
    <title>The page Title</title>
</head>
<body>
    <div class="foo">hello world</div>
</body>
</html>`);


const util = require('util');
console.log('docAST:', util.inspect(docAST, true, 10, true))

```

## Licence

The [MIT](http://opensource.org/licenses/MIT) License

Copyright 2017 (c) Gilles Coomans

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the 'Software'), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.6.0"></a>
# [1.6.0](https://github.com/nomocas/elenpi-html-parser/compare/v1.5.1...v1.6.0) (2018-02-28)


### Features

* better management of innerscript coments + add tests ([9d756a5](https://github.com/nomocas/elenpi-html-parser/commit/9d756a5))



<a name="1.5.1"></a>
## [1.5.1](https://github.com/nomocas/elenpi-html-parser/compare/v1.5.0...v1.5.1) (2018-02-07)


### Bug Fixes

* **inner-script:** manage single double-quote or single single-quote in raw content ([6678e64](https://github.com/nomocas/elenpi-html-parser/commit/6678e64))
* **tag-end:** produce real error when tag is baldy closed ([57fd4c8](https://github.com/nomocas/elenpi-html-parser/commit/57fd4c8))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/nomocas/elenpi-html-parser/compare/v1.4.0...v1.5.0) (2018-01-25)


### Features

* add namespaced attribute's name ([021eb7e](https://github.com/nomocas/elenpi-html-parser/commit/021eb7e))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/nomocas/elenpi-html-parser/compare/v1.3.0...v1.4.0) (2018-01-19)


### Features

* add handlebars variables forwarding for partials attributes ([bebce2b](https://github.com/nomocas/elenpi-html-parser/commit/bebce2b))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/nomocas/elenpi-html-parser/compare/v1.2.0...v1.3.0) (2018-01-09)


### Features

* allow dot syntax for attributes ([15d3db3](https://github.com/nomocas/elenpi-html-parser/commit/15d3db3))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/nomocas/elenpi-html-parser/compare/v1.1.0...v1.2.0) (2018-01-05)


### Features

* add handlebars attributes parsing ([5b4c49d](https://github.com/nomocas/elenpi-html-parser/commit/5b4c49d))



<a name="1.1.0"></a>
# [1.1.0](https://github.com/nomocas/elenpi-html-parser/compare/v1.0.4...v1.1.0) (2017-11-23)


### Features

* **content-position:** add content-position for all tags ([ab788a7](https://github.com/nomocas/elenpi-html-parser/commit/ab788a7))



<a name="1.0.4"></a>
## [1.0.4](https://github.com/nomocas/elenpi-html-parser/compare/v1.0.3...v1.0.4) (2017-11-20)


### Bug Fixes

* **comments:** fix comments with special chars ([1685a5c](https://github.com/nomocas/elenpi-html-parser/commit/1685a5c))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/nomocas/elenpi-html-parser/compare/v1.0.2...v1.0.3) (2017-11-20)


### Bug Fixes

* **comments:** fix nested comments parsing and add related test ([651aeed](https://github.com/nomocas/elenpi-html-parser/commit/651aeed))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/nomocas/elenpi-html-parser/compare/v1.0.1...v1.0.2) (2017-11-15)


### Bug Fixes

* **all:** follow hbs needs ([008ddb0](https://github.com/nomocas/elenpi-html-parser/commit/008ddb0))
* **ast:** start changing children to childNodes ([c111c78](https://github.com/nomocas/elenpi-html-parser/commit/c111c78))
* **test:** follow last modification ([bfe9c9a](https://github.com/nomocas/elenpi-html-parser/commit/bfe9c9a))



<a name="1.0.1"></a>
## 1.0.1 (2017-11-14)


### Bug Fixes

* **base:** finalise extraction and first tests ([d27d7eb](https://github.com/nomocas/elenpi-html-parser/commit/d27d7eb))
* **dependency:** update elenpi version to 0.5.1 ([f4524d0](https://github.com/nomocas/elenpi-html-parser/commit/f4524d0))


test/test.js
/* global describe, it */
import chai from 'chai';

const expect = chai.expect; 

describe('Your tests', () => {
	it('dummy', () => {
		expect(true).equal(true);
	});
});


test/html-parser.spec.js
/**
 * @author Gilles Coomans <gilles.coomans@gmail.com>
 *
 */

/* global describe, it */
import chai from 'chai';
import parser from '../src/index';

const expect = chai.expect;
chai.should();

describe("HTML5 parse", () => {

	describe("tag", () => {
		const res = parser.parse('<div></div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div"
			});
		});
	});

	describe("self closed tag", () => {
		const res = parser.parse('<div/>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div"
			});
		});
	});

	describe("unstrict self-closing tag", () => {
		const res = parser.parse('<br>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "br"
			});
		});
	});

	describe("open tag", () => {
		const res = parser.parse('<meta>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "meta"
			});
		});
	});

	describe("textWithoutQuotesOrTags 1", () => {
		const res = parser.parse('ho', 'textWithoutQuotesOrTags');
		it("should", () => {
			expect(res).to.deep.equals({
				content: 'ho'
			});
		});
	});


	describe("textWithoutQuotesOrTags 2", () => {
		const res = parser.parse('ho < bar', 'textWithoutQuotesOrTags');
		it("should", () => {
			expect(res).to.deep.equals({
				content: 'ho < bar'
			});
		});
	});

	describe("textWithoutQuotesOrTags 3", () => {
		const res = parser.parse('ho<bar/>', 'fragment');
		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "#text",
					value: "ho"
				}, {
					nodeName: "bar"
				}]
			});
		});
	});

	describe("textWithoutQuotesOrTags 4", () => {
		const res = parser.parse('<title>ho</title>', 'fragment');
		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					childNodes: [{
						nodeName: "#text",
						value: "ho"
					}],
					nodeName: "title"
				}]
			});
		});
	});

	describe("open tag + closed just after", () => {
		const res = parser.parse('<meta><title>ho</title>', 'fragment');
		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "meta"
				}, {
					childNodes: [{
						nodeName: '#text',
						value: "ho"
					}],
					nodeName: "title"
				}]
			});
		});
	});

	describe("tag badly closed will throw", () => {
		const parse = function() { parser.parse('<bars>hello</bar>', 'tag'); };
		it("should throw", () => {
			expect(parse).to.throw();
		});
	});


	describe("tag with text content", () => {
		const res = parser.parse('<div>hello</div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div",
				childNodes: [{
					nodeName: '#text',
					value: "hello"
				}]
			});
		});
	});

	describe("attribute (double quoted)", () => {
		const res = parser.parse('class="hello"', 'attribute');
		it("should", () => {
			expect(res).to.deep.equals({
				name: "class",
				value: "hello"
			});
		});
	});

	describe("attribute (single quoted)", () => {
		const res = parser.parse('class=\'hello\'', 'attribute');
		it("should", () => {
			expect(res).to.deep.equals({
				name: "class",
				value: "hello"
			});
		});
	});

	describe("attribute (directValue)", () => {
		const res = parser.parse('class=hello', 'attribute');
		it("should", () => {
			expect(res).to.deep.equals({
				name: "class",
				value: "hello",
				directValue: true
			});
		});
	});

	describe("tag (self-closed) with attribute (directValue) that touch the end of tag", () => {
		const res = parser.parse('<div class=hello/>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div",
				attributes: [{
					name: "class",
					value: "hello",
					directValue: true
				}]
			});
		});
	});

	describe("attribute with directValue as path", () => {
		const res = parser.parse('class=./hello/foo', 'attribute');
		it("should", () => {
			expect(res).to.deep.equals({
				name: "class",
				value: "./hello/foo",
				directValue: true
			});
		});
	});

	describe("tag attribute with directValue as path", () => {
		const res = parser.parse('<div class=./hello/foo />', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div",
				attributes: [{
					name: "class",
					value: "./hello/foo",
					directValue: true
				}]
			});
		});
	});

	describe("attributes", () => {
		const res = parser.parse('class="hello" id=reu', 'attributes');
		it("should", () => {
			expect(res).to.deep.equals({
				attributes: [{
					name: "class",
					value: "hello"
				}, {
					name: "id",
					value: "reu",
					directValue: true
				}],
			});
		});
	});

	describe("tag with children", () => {
		const res = parser.parse('<div>hello <span>John</span></div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div",
				childNodes: [{
					nodeName: '#text',
					value: "hello "
				}, {
					nodeName: "span",
					childNodes: [{
						nodeName: '#text',
						value: "John"
					}]
				}]
			});
		});
	});

	describe("tag with children with new line and tab", () => {
		const res = parser.parse('<div>hello \n\t<span>John</span>\n</div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div",
				childNodes: [{
					nodeName: '#text',
					value: "hello \n\t"
				}, {
					nodeName: "span",
					childNodes: [{
						nodeName: '#text',
						value: "John"
					}]
				},
				{
					nodeName: "#text",
					value: "\n"
				}
				]
			});
		});
	});

	describe("tag with script", () => {
		const res = parser.parse('<div>hello <script>var a = 12 < 15;</script></div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div",
				childNodes: [{
					nodeName: '#text',
					value: "hello "
				}, {
					nodeName: "script",
					content: "var a = 12 < 15;"
				}]
			});
		});
	});

	describe("tag with script with new line", () => {
		const res = parser.parse('<div>hello <script>var a = 12 < 15;\nvar b = 0; </script></div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "div",
				childNodes: [{
					nodeName: '#text',
					value: "hello "
				}, {
					nodeName: "script",
					content: "var a = 12 < 15;\nvar b = 0; "
				}]
			});
		});
	});

	describe("p tag", () => {
		const res = parser.parse('<p>hello</p>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: "p",
				childNodes: [{
					nodeName: '#text',
					value: "hello"
				}]
			});
		});
	});

	describe("comment", () => {
		const res = parser.parse('<!-- bloupi -->', 'comment');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: '#comment',
				data: " bloupi "
			});
		});
	});

	describe("comment with single quote", () => {
		const res = parser.parse("<!-- don't -->", 'comment');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: '#comment',
				data: " don't "
			});
		});
	});

	describe("comment with new line", () => {
		const res = parser.parse('<!-- \rbloupi\n -->', 'comment');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: '#comment',
				data: ' \rbloupi\n '
			});
		});
	});

	describe("comment with new line", () => {
		const res = parser.parse('<!-- should include critical CSS (https://github.com/addyosmani/critical) -->', 'comment');
		it("should", () => {
			expect(res).to.deep.equals({
				nodeName: '#comment',
				data: ' should include critical CSS (https://github.com/addyosmani/critical) '
			});
		});
	});

	describe("variable with handlebars syntax", () => {
		const res = parser.parse('<div {{ foo }}>bar</div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				attributes: [{
					handlebars: true,
					expression: "{{ foo }}"
				}],
				childNodes: [{
					nodeName: "#text",
					value: "bar"
				}],
				nodeName: "div"
			});
		});
	});

	describe("variable with dot syntax", () => {
		const res = parser.parse('<div foo.bar>zoo</div>', 'tag');
		it("should", () => {
			expect(res).to.deep.equals({
				attributes: [{
					name: "foo.bar"
				}],
				childNodes: [{
					nodeName: "#text",
					value: "zoo"
				}],
				nodeName: "div"
			});
		});
	});


	describe("Scripts", () => {
		const doc2 = `<script>
		document.write('<script>bouh</script>');
		document.write('<script>"bouh\\'</script>');
		document.write("<script>bouh</script>");
		document.write("<script>'bouh\\"</script>");
		</script><script></script>`;
		const res = parser.parse(doc2, 'fragment');

		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "script",
					content: "\n\t\tdocument.write('<script>bouh</script>');\n\t\tdocument.write('<script>\"bouh\\'</script>');\n\t\tdocument.write(\"<script>bouh</script>\");\n\t\tdocument.write(\"<script>'bouh\\\"</script>\");\n\t\t"
				},
				{
					nodeName: "script",
					content: ""
				}
				]
			});
		});
	});


	describe("Scripts with comment with unique single quote", () => {
		const doc2 = `<script>
		var a = 2; // this doesn't work
		</script>`;
		const res = parser.parse(doc2, 'fragment');

		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "script",
					content: "\n\t\tvar a = 2; // this doesn't work\n\t\t"
				}
				]
			});
		});
	});

	describe("Scripts with tags and comment with unique single quote", () => {
		const doc2 = `<script>
		// we add it late because it's 150k to load
		var b = '<div></div>';
		</script>`;
		const res = parser.parse(doc2, 'fragment');

		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "script",
					content: "\n\t\t// we add it late because it's 150k to load\n\t\tvar b = '<div></div>';\n\t\t"
				}
				]
			});
		});
	});


	describe("Style", () => {
		const doc2 = `<style>
			#fixture {
				position: absolute;
				top: -9999;
				left: -9999;
			}
		</style>`;
		const res = parser.parse(doc2, 'fragment');

		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "style",
					content: "\n\t\t\t#fixture {\n\t\t\t\tposition: absolute;\n\t\t\t\ttop: -9999;\n\t\t\t\tleft: -9999;\n\t\t\t}\n\t\t"
				}]
			});
		});
	});

	describe("comment with nested comment", () => {
		const res = parser.parse(`
	<!-- hop --><template><!-- hip --></template>
`, 'fragment');

		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: '#text',
					value: '\n\t'
				},
				{
					nodeName: '#comment',
					data: ' hop '
				},
				{
					nodeName: 'template',
					childNodes: [{
						nodeName: '#comment',
						data: ' hip '
					}]
				},
				{
					nodeName: '#text',
					value: '\n'
				}
				]
			});
		});
	});

	describe("full line", () => {
		const text = '<div id="hello" class=reu >foo <br> <!-- hello \n--> <span class="blu" > bar </span></div><home /> hello <script type="text/javascript">var a = 12, \nb = a < 14;</script>';
		const res = parser.parse(text, 'fragment');

		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "div",
					attributes: [{
						name: "id",
						value: "hello"
					}, {
						name: "class",
						value: "reu",
						directValue: true
					}],
					childNodes: [{
						nodeName: '#text',
						value: "foo "
					}, {
						nodeName: "br"
					}, {
						nodeName: "#text",
						value: " "
					}, {
						nodeName: '#comment',
						data: " hello \n"
					}, {
						nodeName: "#text",
						value: " "
					}, {
						nodeName: "span",
						attributes: [{
							name: "class",
							value: "blu"
						}],
						childNodes: [{
							nodeName: '#text',
							value: " bar "
						}]
					}]
				}, {
					nodeName: "home"
				}, {
					nodeName: '#text',
					value: " hello "
				}, {
					nodeName: "script",
					attributes: [{
						name: "type",
						value: "text/javascript"
					}],
					content: "var a = 12, \nb = a < 14;"
				}]
			});
		});
	});

	describe("fragment with style followed by script", () => {
		const doc = `<style></style><script>mocha.setup("bdd");</script>`;
		const res = parser.parse(doc, 'fragment');
		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					content: "",
					nodeName: "style"
				}, {
					content: "mocha.setup(\"bdd\");",
					nodeName: "script"
				}]
			});
		});
	});

	describe("document", () => {
		// from elenpi/index.html (mocha tests index for browser)
		const doc = `<!DOCTYPE html>
<html>

	<head>
		<meta charset="utf-8">
		<title>elenpi mocha tests</title>
		<link rel="stylesheet" href="./test/mocha.css" />
		<style>
			#fixture {
				position: absolute;
				top: -9999;
				left: -9999;
			}
		</style>
		<script type="text/javascript" src="./index.js"></script>
		<script src="./test/chai.js"></script>
		<script src="./test/mocha.js"></script>
		<script>
			mocha.setup("bdd");
		</script>
		<script src="./test/test.js"></script>
		<script>
			window.onload = function() {
				mocha.run()
			};
		</script>
	</head>

	<body>
		<h2 style="margin-left:30px;"><a href="https://github.com/nomocas/elenpi">elenpi</a> tests</h2>
		<div id="mocha"></div>
	</body>
</html>`;

		const res = parser.parse(doc, 'document');

		it("should", () => {
			expect(res).to.deep.equals({
				childNodes: [{
					nodeName: "html",
					childNodes: [{
						nodeName: "#text",
						value: "\n\n\t"
					},
					{
						nodeName: "head",
						childNodes: [{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "meta",
							attributes: [{
								name: "charset",
								value: "utf-8"
							}]
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "title",
							childNodes: [{
								nodeName: "#text",
								value: "elenpi mocha tests"
							}]
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "link",
							attributes: [{
								name: "rel",
								value: "stylesheet"
							},
							{
								name: "href",
								value: "./test/mocha.css"
							}
							]
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "style",
							content: "\n\t\t\t#fixture {\n\t\t\t\tposition: absolute;\n\t\t\t\ttop: -9999;\n\t\t\t\tleft: -9999;\n\t\t\t}\n\t\t"
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "script",
							attributes: [{
								name: "type",
								value: "text/javascript"
							},
							{
								name: "src",
								value: "./index.js"
							}
							],
							content: ""
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "script",
							attributes: [{
								name: "src",
								value: "./test/chai.js"
							}],
							content: ""
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "script",
							attributes: [{
								name: "src",
								value: "./test/mocha.js"
							}],
							content: ""
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "script",
							content: "\n\t\t\tmocha.setup(\"bdd\");\n\t\t"
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "script",
							attributes: [{
								name: "src",
								value: "./test/test.js"
							}],
							content: ""
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "script",
							content: "\n\t\t\twindow.onload = function() {\n\t\t\t\tmocha.run()\n\t\t\t};\n\t\t"
						},
						{
							nodeName: "#text",
							value: "\n\t"
						}
						]
					},
					{
						nodeName: "#text",
						value: "\n\n\t"
					},
					{
						nodeName: "body",
						childNodes: [{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "h2",
							attributes: [{
								name: "style",
								value: "margin-left:30px;"
							}],
							childNodes: [{
								nodeName: "a",
								attributes: [{
									name: "href",
									value: "https://github.com/nomocas/elenpi"
								}],
								childNodes: [{
									nodeName: "#text",
									value: "elenpi"
								}]
							},
							{
								nodeName: "#text",
								value: " tests"
							}
							]
						},
						{
							nodeName: "#text",
							value: "\n\t\t"
						},
						{
							nodeName: "div",
							attributes: [{
								name: "id",
								value: "mocha"
							}]
						},
						{
							nodeName: "#text",
							value: "\n\t"
						}
						]
					},
					{
						nodeName: "#text",
						value: "\n"
					}
					]
				}]
			});
		});
	});
});


src/index.js
/** 
 * @author Gilles Coomans <gilles.coomans@gmail.com>
 * html tokenizer
 */

/* eslint no-console:0 */
import elenpi from 'elenpi';

const r = elenpi.Rule.initializer,
	Parser = elenpi.Parser,
	exec = Parser.exec,
	openTags = /^(br|input|area|base|col|command|embed|hr|img|link|meta|param|source|track|wbr)/,
	rawContentTags = /^(?:script|style|code)/,
	doubleString = /^"((?:[^"\\]|\\.)*)"/,
	singleString = /^'((?:[^'\\]|\\.)*)'/;

const rules = {
	document: r
		.zeroOrMore(r.space().one('comment'))
		.terminal(/^\s*<!DOCTYPE[^>]*>\s*/i)
		.one('fragment')
		.space(),

	comment: r.terminal(/^<!--([.\s\S\r]*?)(?=-->)-->/, (env, obj, cap) => {
		obj.nodeName = '#comment';
		obj.data = cap[1];
	}),

	// closing tag
	tagEnd: r.terminal(/^\s*<\/([\w-_\:]+)\s*>/, (env, obj, cap, startIndex) => {
		if (obj.nodeName !== cap[1]) {
			env.error = true;
			env.errorMessage = 'tag badly closed : ' + cap[1] + ' - (at opening : ' + obj.nodeName + ')';
		} else if (env.options && env.options.location)
			obj.endContentIndex = startIndex;
	}),

	// fragment (or children)
	fragment: r
		.zeroOrMore({
			pushTo: 'childNodes',
			rule: r.oneOf('comment', 'text', 'tag')
		}),

	text: r.terminal(/^[^<]+/, (env, obj, cap) => {
		obj.nodeName = '#text';
		obj.value = cap[0];
	}),

	// normal tag (including raw tags)
	tag: r
		.terminal(/^<([\w-_:]+)\s*/, (env, obj, cap) => {
			obj.nodeName = cap[1];
		}) // start tag
		.one('attributes')
		.oneOf(
			// strict self closed tag			
			r.terminal(/^\/>/),

			r.char('>') // open tag or tag with children
			.done((env, obj, lastIndex) => {

				// check html5 unstrict self-closing tags
				if (openTags.test(obj.nodeName))
					return; // no children

				if (env.options && env.options.location) {
					obj.startContentIndex = lastIndex;
				}
				if (rawContentTags.test(obj.nodeName)) {
					// get inner script content
					obj.content = '';
					exec(env.parser.rules.innerScript, obj, env);
				} else
					// get inner tag content
					exec(env.parser.rules.fragment, obj, env);

				if (!env.error) // close tag
					exec(env.parser.rules.tagEnd, obj, env);
			}),
			r.error('Missing end of tag')
		),


	attributes: r.zeroOrMore({
		rule: r.oneOf('handlebars', 'attribute').space(),
		pushTo: (env, parent, descriptor) => {
			parent.attributes = parent.attributes || [];
			if (descriptor.name === 'class' && !descriptor.value)
				return;
			parent.attributes.push(descriptor);
		}
	}),

	attribute: r
		.terminal(/^([\w-_\.:]+)\s*/, (env, obj, cap) => {
			obj.name = cap[1];
		})
		.maybeOne('attributeValue'),

	attributeValue: r
		.terminal(/^\s*=\s*/)
		.oneOf({
			rules: [
				r.terminal(doubleString, (env, obj, cap) => {
					obj.value = cap[1];
				}),
				r.terminal(singleString, (env, obj, cap) => {
					obj.value = cap[1];
				}),
				r.terminal(/^(\/(?!>)|[^\s\/])+/, (env, obj, cap) => {
					obj.value = cap[0];
					obj.directValue = true;
				})
			]
		}),


	handlebars: r
		.terminal(/^({{{?[^}]+}}}?)\s*/, (env, obj, cap) => {
			obj.handlebars = true;
			obj.expression = cap[1];
		}),

	innerScript: r
		.zeroOrMore(
			r.oneOf(
				'textWithoutQuotesOrTagsOrComments',
				'jsComment',
				'jsComment2',
				'doublestring', 
				'singlestring', 
				'templatestring',
				r.terminal(/^['"`\/]/, (env, descriptor, cap) => descriptor.content += cap[0])
			)
		),

	textWithoutQuotesOrTagsOrComments: r
		.terminal(/^([^'"`<\/]|<(?!\/)|\/(?![\/\*]))+/, (env, obj, cap) => {
			if (obj.content)
				obj.content += cap[0];
			else
				obj.content = cap[0];
		}),

	jsComment: r
		.terminal(/^\/\/[^\n]*/, (env, obj, cap) => {
			if (obj.content)
				obj.content += cap[0];
			else
				obj.content = cap[0];
		}),

	jsComment2: r
		.terminal(/^\/\*.*(?=\*\/)/, (env, obj, cap) => {
			if (obj.content)
				obj.content += cap[0];
			else
				obj.content = cap[0];
		}),

	textWithoutQuotesOrTags: r
		.terminal(/^([^'"`<]|<(?!\/))+/, (env, obj, cap) => {
			if (obj.content)
				obj.content += cap[0];
			else
				obj.content = cap[0];
		}),

	doublestring: r.terminal(doubleString, (env, descriptor, cap) => descriptor.content += `"${ cap[1] }"`),

	singlestring: r.terminal(singleString, (env, descriptor, cap) => descriptor.content += `'${ cap[1] }'`),

	templatestring: r.terminal(/^`([^`]*)`/, (env, descriptor, cap) => descriptor.content += `\`${ cap[1] }\``)
};

const parser = new Parser(rules, 'fragment');

parser.parseDocument = function parseDoc(doc, options = {}) {
	return this.parse(doc, 'document', {}, {
		options
	});
};

parser.parseFragment = function parseFrag(doc, options = {}) {
	return this.parse(doc, 'fragment', {}, {
		options
	});
};

export default parser;


karma.conf.js
// Karma configuration
const babel = require('rollup-plugin-babel'),
	babelrc = require('babelrc-rollup').default;

module.exports = function (config) {
	config.set({

		// base path that will be used to resolve all patterns (eg. files, exclude)
		basePath: '',


		// frameworks to use
		// available frameworks: https://npmjs.org/browse/keyword/karma-adapter
		frameworks: ['mocha', 'chai'],


		// list of files / patterns to load in the browser
		files: [{
			pattern: 'src/**/*.js',
			included: false
		}, // watch them but do not include them
			'test/**/*.js'
		],


		// list of files to exclude
		exclude: [],


		// preprocess matching files before serving them to the browser
		// available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
		preprocessors: {
			'src/**/*.js': ['rollup'],
			'test/**/*.js': ['rollup']
		},
		rollupPreprocessor: {
			plugins: [
				babel(babelrc())
			],
			moduleName: 'KarmaBundle',
			format: 'iife', // helps prevent naming collisions
			sourceMap: 'inline', // sensible for testing
		},

		// test results reporter to use
		// possible values: 'dots', 'progress'
		// available reporters: https://npmjs.org/browse/keyword/karma-reporter
		reporters: ['progress'],


		// web server port
		port: 9876,


		// enable / disable colors in the output (reporters and logs)
		colors: true,


		// level of logging
		// possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
		logLevel: config.LOG_INFO,


		// enable / disable watching file and executing tests whenever any file changes
		autoWatch: true,


		// start these browsers
		// available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
		browsers: ['Chrome'],


		// Continuous Integration mode
		// if true, Karma captures browsers, runs the tests and exits
		singleRun: false,

		// Concurrency level
		// how many browser should be started simultaneous
		concurrency: Infinity
	});
};


html-example.js
/* eslint no-console:0 */
const parser = require('./dist/bundles/index');
const util = require('util');

const doc = `<!DOCTYPE html>
<html>
	<head>
		<meta charset="utf-8">
		<title>elenpi mocha tests</title>
		<link rel="stylesheet" href="./test/mocha.css" />
		<style>
			#fixture {
				position: absolute;
				top: -9999;
				left: -9999;
			}
		</style>
		<script type="text/javascript" src="./index.js"></script>
		<script src="./test/chai.js"></script>
		<script src="./test/mocha.js"></script>
		<script>
			mocha.setup("bdd");
		</script>
		<script src="./test/test.js"></script>
		<script>
			window.onload = function() {
				mocha.run()
			};
		</script>
		<script>
			document.write('<script>bouh</script>');
		</script>
	</head>

	<body>
		<h2 style="margin-left:30px;"><a href="https://github.com/nomocas/elenpi">elenpi</a> tests</h2>
		<div id="mocha"></div>
	</body>
</html>`;

const doc2 = `<script>
		document.write('<script>bouh</script>');
		document.write('<script>"bouh\\'</script>');
		document.write("<script>bouh</script>");
		document.write("<script>'bouh\\"</script>");
		</script><script></script>`;


const doc3 = '<script>hello world</script>';

const doc4 = `
<script>hello world</script>
`;

// console.log('doc3 trimed |%s|', doc3.substring(8, 19));
// console.log('doc4 trimed |%s|', doc4.substring(9, 20));

// const res = parser.parse(doc, 'document');
const res2 = parser.parse(doc2, 'children');
// const res = parser.parse(doc3, 'children');
// const res = parser.parse(doc4, 'children');

// console.log('res', util.inspect(res, true, 7, true));
// console.log('res2', util.inspect(res2, true, 7, true));
console.log('res2', JSON.stringify(res2, null, ' '));
